minio/select_object.py

```diff
@@ -0,0 +1,219 @@
+            if byte_int(headers_byte_length) > 0:
+                # Process header and populate header map
+                header, bytes_parsed, chunked_message =  \
+                    extract_component(response, read_buffer,
+                                      bytes_parsed,
+                                      byte_int(headers_byte_length),
+                                      chunked_message)
+                header_map = \
+                extract_header(header, byte_int(headers_byte_length))
+
+                # Process Payload
+                payload_length = byte_int(total_byte_length) \
+                                - byte_int(headers_byte_length) - int(16)
+                payload, bytes_parsed, chunked_message = \
+                    extract_component(response, read_buffer,
+                                      bytes_parsed, payload_length,
+                                      chunked_message)
+
+                if header_map["message-type"] == 'event':
+                    # Parse message only when event-type is Records, 
+                    # Progress, Stats. Break the loop if event type is End
+                    # Do nothing if event type is Cont
+                    if header_map["event-type"] == 'Records' or \
+                        header_map["event-type"] == 'Progress' or \
+                        header_map["event-type"] == 'Stats':
+                        parse_message(header_map, payload, payload_length, record, stats)
+
+                    if header_map["event-type"] == 'End':
+                        break
+                if header_map["message-type"] == 'error':
+                    parse_message(header_map, payload, payload_length, record, stats)
+                    break
+                # Fetch message CRC
+                message_crc, bytes_parsed, chunked_message = \
+                    extract_component(response, read_buffer,
+                                      bytes_parsed, CONS_READ_SIZE,
+                                      chunked_message)
+
+                # Generate complete message
+                complete_message = total_byte_length + \
+                                    headers_byte_length + prelude_crc + \
+                                    header + payload
+                if not validate_crc(complete_message, message_crc):
+                    raise CRCValidationError(
+                        {"Checksum Mismatch, MessageCRC of " + \
+                            str(calcuate_crc(complete_message)) + \
+                            " does not equal expected CRC of "+ \
+                            str(byte_int(message_crc))})
+        else:
+            raise CRCValidationError(
+                {"Checksum Mismatch, MessageCRC of " + \
+                str(calcuate_crc(total_byte_length + headers_byte_length)) + \
+                " does not equal expected CRC of "+ str(byte_int(prelude_crc))})
+    return record.getvalue(), stats.getvalue()
```

Comment:
This shouldn't return multiple values `extract_message` should return io.IOBase compatible reader such that one can read from `for data in r.read(buffer):` can be a stream. 

This will use all the memory on the client and fail subsequently.  The code here extract_message should yield bytes.

Reply:
You would need to implement something like this

```
class SelectObjectReader(object):    
    def __init__(self, response):
        self.response = response

    def readable(self):
        return True

    def writeable(self):
        return False

    @property
    def closed(self):
        return self.response.isclosed()

    def close(self):
        self.response.close()

    def stats(self):
        return self.stats
  
    def progress(self):
        return self.progress

    def read(self, num_bytes):
       ... extract each record from the response body ... and buffer it.
       ... send only up to requested bytes such as message[:num_bytes]
       ... rest is buffered and added to the next iteration.
```

> NOTE: self.stats and self.progress should be parsed already, we shouldn't return direct XML blobs to the user.  Implementing this allows it.